src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceModelHttpMessageHandler.cs

```diff
@@ -27,5 +29,13 @@ public CookieContainer CookieContainer
             get { return _innerHandler.CookieContainer; }
             set { _innerHandler.CookieContainer = value; }
         }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var content = request.Content as MessageContent;
+            var responseMessage = await base.SendAsync(request, cancellationToken);
+            await content.WriteCompletionTask;
```

Comment:
We have a helper in our TaskHelpers class which provides a custom awaitable object which checks for a non-default SyncronizationContext or a non-default TaskScheduler and if it finds either causes the continuation to be run with Task.Run(). As we're always guaranteed to go async at some point as we always need to wait for the response from the server, we decided it was easier and safer to jump threads to a safe thread early on than if we needed than to worry about making sure we don't miss a place where we need to use ConfigureAwait(false). The code to do this is [here](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/src/Internals/System/Runtime/TaskHelpers.cs#L274). We have tests which verify this by using a custom SynchronizationContext which uses a single thread to execute all work and verifies we don't get deadlocks.